runtime/runtime-standalone/src/lib.rs

```diff
@@ -0,0 +1,434 @@
+            epoch_height: 0,
+            block_timestamp: genesis_config.genesis_time,
+            gas_price: genesis_config.gas_price,
+            gas_limit: genesis_config.gas_limit,
+            gas_burnt: 0,
+        }
+    }
+
+    pub fn produce(&self, new_state_root: CryptoHash, epoch_length: u64) -> Block {
+        Self {
+            gas_price: self.gas_price,
+            gas_limit: self.gas_limit,
+            block_timestamp: self.block_timestamp + 1,
+            prev_block: Some(Box::new(self.clone())),
+            state_root: new_state_root,
+            block_height: self.block_height + 1,
+            epoch_height: (self.block_height + 1) / epoch_length,
+            gas_burnt: 0,
+        }
+    }
+}
+
+pub struct RuntimeStandalone {
+    genesis: GenesisConfig,
+    tx_pool: TransactionPool,
+    transactions: HashMap<CryptoHash, SignedTransaction>,
+    outcomes: HashMap<CryptoHash, ExecutionOutcome>,
+    cur_block: Block,
+    runtime: Runtime,
+    trie: Arc<Trie>,
+    pending_receipts: Vec<Receipt>,
+}
+
+impl RuntimeStandalone {
+    pub fn new(genesis: GenesisConfig, store: Arc<Store>) -> Self {
+        let mut genesis_block = Block::genesis(&genesis);
+        let mut store_update = store.store_update();
+        let runtime = Runtime::new(genesis.runtime_config.clone());
+        let trie = Arc::new(Trie::new(store));
+        let trie_update = TrieUpdate::new(trie.clone(), CryptoHash::default());
+        let (s_update, state_root) =
+            runtime.apply_genesis_state(trie_update, &[], &genesis.state_records);
+        store_update.merge(s_update);
+        store_update.commit().unwrap();
+        genesis_block.state_root = state_root;
+        Self {
+            genesis,
+            trie,
+            runtime,
+            transactions: HashMap::new(),
+            outcomes: HashMap::new(),
+            cur_block: genesis_block,
+            tx_pool: TransactionPool::new(),
+            pending_receipts: vec![],
+        }
+    }
+
+    pub fn new_with_store(genesis: GenesisConfig) -> Self {
+        RuntimeStandalone::new(genesis, create_test_store())
+    }
+
+    /// Processes blocks until the final value is produced
+    pub fn resolve_tx(
+        &mut self,
+        mut tx: SignedTransaction,
+    ) -> Result<ExecutionOutcome, RuntimeError> {
+        tx.init();
+        let mut outcome_hash = tx.get_hash();
+        self.transactions.insert(outcome_hash, tx.clone());
+        self.tx_pool.insert_transaction(tx);
+        loop {
+            self.process_block()?;
+            let outcome = self.outcomes.get(&outcome_hash).unwrap();
+            match outcome.status {
+                ExecutionStatus::Unknown => unreachable!(), // ExecutionStatus::Unknown is not relevant for a standalone runtime
+                ExecutionStatus::SuccessReceiptId(ref id) => outcome_hash = *id,
+                ExecutionStatus::SuccessValue(_) | ExecutionStatus::Failure(_) => {
```

Comment:
Why return only an `Ok`?  Why not return an `Err` on a Failure?

Reply:
Hey, actually it’s a good point.. Would be nice to return Err. 
https://github.com/nearprotocol/nearcore/pull/2598#discussion_r422351327

The problem is that ExecutionStatus::SuccessValue or ExecutionStatus::Failure represents a semantics of on a level of a smart contract. For example, RuntimeError can be only `InvalidTxError`, but not `ActionError`.. 

https://github.com/nearprotocol/nearcore/blob/090635875137f31de81561d810e505a4699e77ae/core/primitives/src/errors.rs#L15-L20

https://github.com/nearprotocol/nearcore/blob/090635875137f31de81561d810e505a4699e77ae/core/primitives/src/errors.rs#L45-L58